torch/nn/modules/container.py

```diff
@@ -187,6 +187,122 @@ def extend(self, modules):
         return self
 
 
+class ModuleDict(Module):
+    r"""Holds submodules in a dictionary.
+
+    ModuleDict can be indexed like a regular Python dictionary, but modules it
+    contains are properly registered, and will be visible by all Module methods.
+
+    Arguments:
+        modules (iterable, optional): a mapping (dictionary) of (string: module)
+            or an iterable of key/value pairs of type (string, module)
+
+    Example::
+
+        class MyModule(nn.Module):
+            def __init__(self):
+                super(MyModule, self).__init__()
+                self.choices = nn.ModuleDict({
+                        'conv': nn.Conv2d(10, 10, 3),
+                        'pool': nn.MaxPool2d(3)
+                })
+                self.activations = nn.ModuleDict([
+                        ['lrelu', nn.LeakyReLU()],
+                        ['prelu', nn.PReLU()]
+                ])
+
+            def forward(self, x, choice, act):
+                x = self.choices[choice](x)
+                x = self.activations[act](x)
+                return x
+    """
+
+    def __init__(self, modules=None):
+        super(ModuleDict, self).__init__()
+        if modules is not None:
+            self.update(modules)
+
+    def __getitem__(self, key):
+        return self._modules[key]
+
+    def __setitem__(self, key, module):
+        self.add_module(key, module)
```

Comment:
Can you add a check in `nn.Module.add_module` that `key` is a string (using `torch._six. string_types`), and also a test for this case? 

Because builtin dict can have arbitrary hashable objects as keys, people may use things other than strings, and we should give them a good error message.

Reply:
> builtin dict can have arbitrary hashable objects as keys

@SsnL I am using `Module.add_module` which does these checks (though very implicitly).

```python
>>> import torch
>>> from torch import nn
>>> class A(nn.Module):
...     def __init__(self):
...             super().__init__()
...             self.add_module(123, nn.ReLU())
...     def forward(self):
...             return 0
... 
>>> A()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 4, in __init__
  File "/home/karan/anaconda3/lib/python3.6/site-packages/torch/nn/modules/module.py", line 166, in add_module
    elif hasattr(self, name) and name not in self._modules:
TypeError: hasattr(): attribute name must be string
```

I will add the check anyway, but it probably needs a cleaner fix in 
 `module.py`

Reply:
lol you are right. it's very implicit indeed. 

please add the check in `module.py`, thank you!

Reply:
Done